plutus-core-interpreter/src/Language/PlutusCore/Interpreter/LMachine.hs

```diff
@@ -0,0 +1,278 @@
+-- argument closure C from the stack, generate a new heap location l,
+-- bind C (unevaluated) to l in the heap, then evaluate e in
+-- env[x->l].  When we come across a use of x inside e, we look up l
+-- in the heap to get N (inside the closure C): if it's Evaluated then
+-- we continue immediately, using N instead of x. Otherwise we place
+-- an update marker on the stack and evaluate N to get a new closure
+-- C' containing a value V (and a possibly new environment): by
+-- this time we should be back at the update marker containing l, and
+-- we store Evaluated C back in the heap at l, then continue
+-- evaluating e.
+
+
+module Language.PlutusCore.Interpreter.LMachine
+    ( EvaluationResultF (EvaluationSuccess, EvaluationFailure)
+    , EvaluationResult
+    , evaluateL
+    , runL
+    ) where
+
+import           Language.PlutusCore
+import           Language.PlutusCore.Constant
+import           Language.PlutusCore.View
+import           PlutusPrelude
+
+import           Control.Monad.Identity
+import           Data.IntMap                  (IntMap)
+import qualified Data.IntMap                  as IntMap
+
+type Plain f = f TyName Name ()
+
+-- Let's just throw exceptions for the time being: error handling is changeable at the moment
+data LMachineException =
+    LMachineException String (Plain Term)
+  | LMachineStringException String
+    deriving (Show, Typeable)
+
+instance Exception LMachineException
+
+-- | A term together with an enviroment mapping free variables to heap locations
+data Closure = Closure
+    { _closureValue       :: Plain Term
+    , _closureEnvironment :: Environment
+    } deriving (Show)
+
+-- | L machine environments
+-- Each entry is a mapping from the 'Unique' representing a variable to a heap location
+newtype Environment = Environment (IntMap HeapLoc)
+    deriving (Show)
+
+-- | Heap location. Int should always be big enough: maxBound::Int is 2^63-1
+type HeapLoc = Int
+
+-- | Heap entries: unevaluated and evaluated lazy function arguments.
+data HeapEntry =
+    Evaluated   Closure  -- The term in this closure should always be a value. It would be good if the type system could distiniguish these.
+  | Unevaluated Closure
+
+-- | The heap itself: a map from heap locations to
+-- unevaluated/evaluated terms, together with an integer counter
+-- @_top@ which serves as the current heap location.
+data Heap = Heap {
+      _heap :: IntMap HeapEntry
+    , _top  :: Int
+}
+
+{- Environments map variable ids to locations, the heap maps locations
+   to terms.  We really do require this indirection: there may be
+   multiple objects in the heap corresponding to the same variable
+   (during recursion, for example).  For the same reason we always
+   work with closures instead of terms; this also makes it easier
+   to ensure that we're using the correct environment. -}
+
+
+-- | Stack frames.  These are effectively continuations: would using explicit CPS speed things up?
```

Comment:
Re: explicit CPS, we need to be able to inspect the structure of the frame, which I think precludes making them actual continuations?

Reply:
I think it's OK (although thinking about continuations too much isn't good for my brain).  The continuation would just "know" what constructor it represents, and would contain any other information it needs in a closure, a bit like the Scott encoding. The `computeL` function would replace the stack/context parameter with a continuation and the `returnL` function would disappear and be replaced by applications of continuations to values.  For example, in the case for `Var` in `computeL`, instead of pushing `FrameUpdateHeap l` onto the stack, you'd use the continuation `\result -> k ctx heap' result where heap' = updateHeap l ...` which includes the thing that `returnL` has to do when it finds an update frame on the stack. Here `k` is the incoming continuation, so the stack gets replaced by a chain of continuations.

I may try to implement this just to make sure that it works.

Reply:
Hm, interesting. I'd say try it out on the CEK machine first, but the implementation of that is monadic, so you might have to use `ConT` which is always a bit mindblowing.